### ISSUE TITLE
Possible unnecessary null tests in ArraySortHelper<TKey,TValue> 

### ISSUE THREAD

**Issue by colgreen** [dotnet/coreclr#14162]
We can pass a secondary array to Array.Sort() like so:

    Array.Sort<int,int>(keys, values, ...)

This overload of Array.Sort() is performed by the ArraySortHelper<TValue,TKey> class rather than `ArraySortHelper<TKey>`. However, `ArraySortHelper<TValue,TKey>` is still performing null tests on the 'values' array at each swap operation, which I think is redundant. 

Performing a conditional branch like this at each swap (of which there may typically be thousands per call to sort()) will have a notable performance impact, so I think this is a good candidate for a performance improvement without any backwards compatibility issues.

**Comment by danmosemsft** [comment]
@colgreen would you like to take a shot at a PR, with some perf numbers?

**Comment by colgreen** [comment]
Yes ok, I will try to find time to do that. Looks like the various TrySZSort() calls handle the 'native' data types anyway, so not sure if the same problem exists within those native/CLR implemented sort implementations. I.e. in order to trigger the above code pathway it looks liek it's necessary to use some user defined type for TKey, so maybe I'll test with both refs and structs of various sizes (say 4, 8, 16 and 32 bytes).

**Comment by colgreen** [comment]
Performance benchmarks.

Each performance test consists of sorting of 1,000 Int32 arrays initialised with uniformly random values;  array length is 100,000. 

All times in the table below are milliseconds.

Note. The tests are performed directly against ArraySortHelper. If Array.Sort is called with an Int32 array then ArraySortHelper is bypassed by TrySZSort(). More benchmark info in next comment.

Host CPU /OS | 32/64 bit CLR | ArraySort | ArraySort (modified) | GenericArraySort | GenericArraySort (modified)
-----------|----------------|-----------|------------------------|--------------------|--------------------------------
Core i7 6700T (Skylake) @ 2.8Ghz / Windows 10 | 64bit | 11.9  | 11.749  | 6.944 | 6.928
Core i7 6700T (Skylake) @ 2.8Ghz / Windows 10 | 32bit | 13.357  | 13.485  | 7.553| 7.442
Core i7 5600U (Broadwell-U) @ 2.6 Ghz | 64bit | 12.658 | 12.456 | 7.07 | 6.89
Core i7 5600U (Broadwell-U) @ 2.6 Ghz | 32bit | 14.764 | 14.488 | 7.567 | 7.586
Core i3-4012Y (Haswell) @ 1.5Ghz / Windows 10 | 64 bit | 26.805 | 26.483 | 15.119 | 14.927
Core i3-4012Y (Haswell) @ 1.5Ghz / Windows 10 | 32 bit | 32.253 | 31.949 | 17.499 | 16.673
Celeron N3150 @ 1.6Ghz / Linux | 64bit | 35.707 | 35.558 | 15.45 | 15.261
Armv7 rev 5 (v7l) raspberry pi2 / Linux | 32bit | 109.561 | 107.764 | 62.997 | 60.593

In summary, the performance improvement is marginal but significant enough to be empirically observable.  The small gains may in part be due to branch prediction logic likely having a near 100% prediction success rate on this code pattern. Note however that the Armv7 CPU does not have branch prediction, although it does have conditional CPU instructions.

**Comment by colgreen** [comment]
For the record here's the benchmark code. 

I normally use BenchmarkDotNet but there are some issues around dotnetcore2 and switc hing between 32/64bit, so I went for this simple benchmark approach, running a release build with no debugger attached.

```
    internal class SimpleBenchmark
    {
        const int __length = 100000;
        const int __loops = 1000;

        Random _rng = new Random(0);
        int[] _keyArr = new int[__length];
        int[] _valArr = new int[__length];
        IArraySortHelper<int,int> _arraySortHelper;

        #region Constructor

        public SimpleBenchmark(IArraySortHelper<int,int> arraySortHelper)
        {
            _arraySortHelper = arraySortHelper;
        }

        #endregion

        #region Public Methods

        public double Run()
        {
            // Warm up phase.
            for(int i=0; i<10; i++)
            {
                Init();
                _arraySortHelper.Sort(_keyArr, _valArr, 0, _keyArr.Length, null);
            }

            // Benchmark phase.
            Stopwatch sw = new Stopwatch();
            for(int i=0; i<__loops; i++)
            {
                Init();
                sw.Start();
                _arraySortHelper.Sort(_keyArr, _valArr, 0, _keyArr.Length, null);
                sw.Stop();
            }

            double msPerSort = (double)sw.ElapsedMilliseconds / __loops;
            return msPerSort;
        }

        #endregion

        #region Private Methods [Init]

        private void Init()
        {
            _keyArr = InitArray(_keyArr);
            _valArr = InitArray(_valArr);
        }

        private int[] InitArray(int[] arr)
        {
            for(int i=0; i<arr.Length; i++) {
                arr[i] = _rng.Next();
            }
            return arr;
        }

        #endregion

        public static void RunTests()
        {
            var benchmark = new SimpleBenchmark(new ArraySortHelper<int, int>());
            double msPerSort = benchmark.Run();
            Console.WriteLine($"ArraySortHelper:\t{msPerSort} ms");

            benchmark = new SimpleBenchmark(new GenericArraySortHelper<int, int>());
            msPerSort = benchmark.Run();
            Console.WriteLine($"GenericArraySortHelper:\t{msPerSort} ms");
            Console.WriteLine("");

            benchmark = new SimpleBenchmark(new ArraySortHelperMod<int, int>());
            msPerSort = benchmark.Run();
            Console.WriteLine($"ArraySortHelperMod:\t{msPerSort} ms");

            benchmark = new SimpleBenchmark(new GenericArraySortHelperMod<int, int>());
            msPerSort = benchmark.Run();
            Console.WriteLine($"GenericArraySortHelperMod:\t{msPerSort} ms");
        }
    }
```

**Comment by danmosemsft** [comment]
@colgreen oops, somehow I missed that you had replied here. This seems a worthwhile change to have in a PR. @jkotas do you know why these null checks existed? (https://github.com/colgreen/coreclr/commit/0f84a622eaf20021a3eacc3c4ee41ccbc517d190)

**Comment by jkotas** [comment]
There was just one `ArraySortHelper<TKey, TValue>` originally that was used to sort both TKey and TKey+TValue cases: https://github.com/emuikernel/SharedSourceCLI/blob/e5abf1c064c9e599a9065c7fbcc82059b5c4755e/sscli20/clr/src/bcl/system/collections/generic/arraysorthelper.cs#L55 It is still like that in CoreRT.

Then somebody created optimized duplicate ArraySortHelper for the TKey case, but forgot to remove the handling of TKey case from the original `ArraySortHelper<TKey, TValue>`.

**Comment by colgreen** [comment]
That was the conclusion I came to too from looking at the code, I couldn't see any code path that would use the null tests so the PR request I submitted should be a safe change (non-breaking change, functionally equivalent, etc.)

**Comment by jkotas** [comment]
Fixed by #14277
